### PR TITLE
xen: modular expression

### PR DESCRIPTION
**Pull request by oxij**
###### Motivation for this change

See commit messages.

###### Things done

- [X] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [X] NixOS
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).
- [X] Works for me.

###### Things not done

- [ ] Tested usability of all qemu binary files
- [ ] Tested all variants to boot

This is pretty much a complete rewrite of the `xen`'s expression.
Test on some your machines before merging!

### CODE REVIEW

**Comment by oxij** [comment]
Rebased on master and added a patch for nixos.

[X] Tested that all predefined xen expressions boot

File: pkgs/applications/virtualization/xen/packages.nix [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        Slimmed-down version of Xen without `qemu-traditional` (you
+        don't need it if you don't know what it is). Use with
+        `qemu_xen-light` from nixpkgs.
+      '';
+    };
+  };
+
+  xen-vanilla = xen_4_5-vanilla;
+  xen-slim = xen_4_5-slim;
+  xen-light = xen_4_5-light;
```
Comment (grahamc) [inline pkgs/applications/virtualization/xen/packages.nix]: Why add a `light` version, if not many people will want it?

**Comment by grahamc** [comment]
Superficially, this looks good. I don't know Xen, nor really understand how its pieces fit together. I just happen to have applied a lot of XSA patches to it. One concern I have is Xen 4.5 seems to be at  the end of its lifecycle. Not yet, but coming up, with 4.6 and 4.7 already out. Can this PR also update to 4.7?

File: pkgs/applications/virtualization/xen/packages.nix (hunk 1 above)
Comment (oxij) [inline pkgs/applications/virtualization/xen/packages.nix]: Why? If you only run Linux VMs or don't have any old Windows VMs that need `qemu-traditional` `light` makes total sense.

**Comment by oxij** [comment]
Not in this PR, please no.

The result would be that this won't ever get applied until 4.7 works as well as 4.5 and I'll have to backport new patches to nixpkgs xen to my branch for another year.

The core of this patch got born quite awhile ago when I tried to update to newly released 4.6 and before old versions were scrubbed from nixpkgs and I backported bleeding from all the orifices four or five times since then.

Let's just merge this so that if anyone gets serious about getting 4.7 working they'll do it on a less copy-pastey and more modular foundation. I can't make 4.6 work well enough yet.

One piece at a time, please.

**Comment by oxij** [comment]
Travis is an idiot. Builds and boots fine on my machines.

**Review by grahamc — COMMENTED** [review]
I'm fairly :thumbsdown: on having the three split-out versions, here
are my reasons:

Overall, as the defacto maintainer of Xen (despite not being a user)
I'm not really excited about having 3 variations of Xen packaged.
Remaining as close as possible to upstream makes my life easier.

Looking at the `longDescriptions` of the three variations, I don't see
good technical justifications for having three versions, other than
build time and space saving. These seems to be fairly weak arguments:

1. Typical VM usage involves lots of wasted space as each VM has
   plenty of its own overhead.
2. Typical users will be receiving a packaged Xen from the upstream
   cache, eliminating build time.
3. Users that don't use the upstream cache are going to be spending
   plenty of time compiling, that a reduction in Xen isn't going to be
   a sizable improvement.
4. Having three will be an _increase_ in resource utilization for
   Hydra, and more personally impactful, a much greater increase in
   time required to test when I apply patches to Xen.
   
Indeed, the `-slim` variation even calls out that its parts may get
out of sync. I don't know what that means or how to fix it, but it
doesn't seem like a good scenario.

This makes me wonder why someone would want to make these changes in
the first place, and wonder if the additional code complexity is worth
the configurability.

If there are more compelling and specific reasons to have the
configurability, I could be convinced that it is worth it.

If there are more compelling and specific reasons to have variations
provided by Nixpkgs, I could be convinced it is worth that, too.

I wouldn't mind these changes so much, but current Xen maintenance
typically falls to me. Having additional configurability makes it
harder to know if my changes are breaking something, and named
variations adds an implicit signal that this configuration is a
supported.

**Comment by oxij** [comment]
> Overall, as the defacto maintainer of Xen (despite not being a user)
> I'm not really excited about having 3 variations of Xen packaged.
> Remaining as close as possible to upstream makes my life easier.

Debian, the rest of nixpkgs and I disagree. You don't compile a separate
`ffmpeg` for every program that uses it, hence you shouldn't compile two
(!) separate `qemu`s for `xen`.

I think its wise to reuse as much nixpkgs as possible. `qemu`'s
expression from nixpkgs has ~6 "*Support" switches that are directly
applicable here. Should I duplicate them in `xen` expression too? That's
nuts.

The problem with `xen` is that its tarballs are too much
"enterprise-level software". Fetching of sources from the net in `make`
is nuts, building multiple internal submodules that are lager than `xen`
itself is nuts too. Yes, its easier to distribute stuff this way in
distribution-agnostic way, but there's no reason to comply with this
enterprise-level craziness with nix. Nix can build stuff by itself.

Ideally, we want to drop `qemu` building from `xen` expression
altogether, override `qemu`'s `src` with corresponding revision from
`xen`'s fork and provide that with vanilla xen. Not sure how this is
gonna work with `qemu-traditional`, probably needs some code stealing
from Debian and/or more options to `qemu` expression. This patch set
moves `xen` in that direction without, AFAICS, loosing anything.

> 1. Typical VM usage involves lots of wasted space as each VM has
>    plenty of its own overhead.

Typical - maybe. But you can use `xen` as a kind of micro-kernel to run
bare-metal programs in OCaml and Haskell, and those don't waste anything
and run on credit-card-sized computers.

> 2. Typical users will be receiving a packaged Xen from the upstream
>    cache, eliminating build time.
> 3. Users that don't use the upstream cache are going to be spending
>    plenty of time compiling, that a reduction in Xen isn't going to be
>    a sizable improvement.

`qemu` built with all the "*Support" options set is 271M. I think that
is considerable. Arguably, you don't need that for typical `xen` usage.
Typical x86-only `qemu` is 26M. Multiply that by two. You can get very
usable suckless.org Linux systems fit in 50M.

> 4. Having three will be an _increase_ in resource utilization for
>    Hydra, and more personally impactful, a much greater increase in
>    time required to test when I apply patches to Xen.

That's backwards logic: "now we have more test cases, so we have more
opportunities to fail them, so lets remove the tests".

> Indeed, the `-slim` variation even calls out that its parts may get
> out of sync. I don't know what that means or how to fix it, but it
> doesn't seem like a good scenario.

`ffmpeg` gets out of sync with its referers too. What do we do in that
case? We make another expression and hold it till its needed or update
outdated packages. I see no problem here.

> This makes me wonder why someone would want to make these changes in
> the first place, and wonder if the additional code complexity is worth
> the configurability.

Complexity where? All options influence is this:

  configureFlags = []
    ++ optional (!withInternalQemu) "--with-system-qemu" # use qemu from PATH
    ++ optional (withInternalTraditionalQemu) "--enable-qemu-traditional"
    ++ optional (!withInternalTraditionalQemu) "--disable-qemu-traditional"

    ++ optional (withSeabios) "--with-system-seabios=${seabios}"
    ++ optional (!withInternalSeabios && !withSeabios) "--disable-seabios"

    ++ optional (withOVMF) "--with-system-ovmf=${OVMF}"
    ++ optional (withInternalOVMF) "--enable-ovmf";

and xenfiles.

Everything else is a refactoring to allow simpler per-version
expressions. 90% of `4.5.nix` is references to sources, not build
scripts. I'd say its less complex now than it was before.

(Also note that OVMF was not supported before, I can't build vanilla
Xen's OVMF. And the external one builds.)

> If there are more compelling and specific reasons to have the
> configurability, I could be convinced that it is worth it.

See above.

> If there are more compelling and specific reasons to have variations
> provided by Nixpkgs, I could be convinced it is worth that, too.

See above.

> I wouldn't mind these changes so much, but current Xen maintenance
> typically falls to me. Having additional configurability makes it
> harder to know if my changes are breaking something, and named

See "backwards logic" above.

> variations adds an implicit signal that this configuration is a
> supported.

That's correct. But the thing is that these configurations are actually
more useful than the vanilla one, unless maybe if you just trying `xen`
out for the first time or something.

But, again, if you don't use it yourself then what's the difference? I
assume you simply compiled it before. Now you can compile three
versions, or, if you're lazy, just continue to compile the vanilla
version, I'll take care of the rest. (Actually, the most likely case is
that if vanilla compiles, the rest will compile too, since they compile
_less_).

Also, if you look at my contribution history to nixpkgs, you'll notice
that changes to `xen` were among the first things I ever contributed
here af32fd6ce3ce2ae138c2fc2bc5c6067216f8b828

I'd say I have enough history with `xen` expression to know what power
users need.

> Sorry, I misread the description and thought you implied you don't need the light version unless you know why you need it, whereas it is clear now you meant you don't need qemu-traditional unless you know why you need it.

Correct. Feel free to reword that to be clearer.

Btw, `./nixos/modules/virtualisation/xen-dom0.nix` is fairly crazy too.
It sets up its own bridges and firewall config, runs its own `dnsmasq`
and stuff. I kinda planned to fix that later too (currently I just
comment out that part in my branch and setup everything myself
per-machine, because who needs options, right?).

In short, for me your comment reads like: "more (not useful for me)
options (I don't use xen, btw) introduce new test cases, meh, don't want
to run those, let's not".

I send the-rays-of-cosmic-love your way for the time you put into
reviewing this PR, but I disagree.

My proposal: read the old expression, read the new expression, ignore
new configurability, feel the difference (I think the level of space
required to correctly indent the expression usually correlates with its
craziness, this patch considerably reduces that metric), then think
about poor power users that want to run MirageOS on their credit-cards
with Xen and want all those options, reconsider.

**Comment by grahamc** [comment]
Jan Malakhovski <notifications@github.com> writes:

> Debian, the rest of nixpkgs and I disagree. You don't compile a separate
> `ffmpeg` for every program that uses it, hence you shouldn't compile two
> (!) separate `qemu`s for `xen`.
>
> I think its wise to reuse as much nixpkgs as possible. `qemu`'s
> expression from nixpkgs has ~6 "*Support" switches that are directly
> applicable here. Should I duplicate them in `xen` expression too? That's
> nuts.

This seems like sound reasoning.

> The problem with `xen` is that its tarballs are too much
> "enterprise-level software". Fetching of sources from the net in `make`
> is nuts, building multiple internal submodules that are lager than `xen`
> itself is nuts too. Yes, its easier to distribute stuff this way in
> distribution-agnostic way, but there's no reason to comply with this
> enterprise-level craziness with nix. Nix can build stuff by itself.

This must already be handled by our current expression, since make can't
access the network anyway.

>> 4. Having three will be an _increase_ in resource utilization for
>>    Hydra, and more personally impactful, a much greater increase in
>>    time required to test when I apply patches to Xen.
>
> That's backwards logic: "now we have more test cases, so we have more
> opportunities to fail them, so lets remove the tests".

No, I disagree. This is adding many more features without test cases.
Many more ways it could build, but produce invalid results.

>> This makes me wonder why someone would want to make these changes in
>> the first place, and wonder if the additional code complexity is worth
>> the configurability.
>
> Complexity where? All options influence is this:

well, that, but I was also surprised to see the over-all nix expression
gain 200 lines of code over the old version. Not that this is
necessarily bad.

> Everything else is a refactoring to allow simpler per-version
> expressions. 90% of `4.5.nix` is references to sources, not build
> scripts. I'd say its less complex now than it was before.
>
> (Also note that OVMF was not supported before, I can't build vanilla
> Xen's OVMF. And the external one builds.)

Nice.

>> I wouldn't mind these changes so much, but current Xen maintenance
>> typically falls to me. Having additional configurability makes it
>> harder to know if my changes are breaking something, and named
>
> See "backwards logic" above.

I disagree it is backwards logic.

>> variations adds an implicit signal that this configuration is a
>> supported.
>
> That's correct. But the thing is that these configurations are actually
> more useful than the vanilla one, unless maybe if you just trying `xen`
> out for the first time or something.
>
> But, again, if you don't use it yourself then what's the difference? I
> assume you simply compiled it before. Now you can compile three
> versions, or, if you're lazy, just continue to compile the vanilla
> version, I'll take care of the rest. (Actually, the most likely case is
> that if vanilla compiles, the rest will compile too, since they compile
> _less_).

I am thrilled to have another Xen maintainer around who I can ping when
issues come up :)

> Also, if you look at my contribution history to nixpkgs, you'll notice
> that changes to `xen` were among the first things I ever contributed
> here af32fd6ce3ce2ae138c2fc2bc5c6067216f8b828

I did notice that, but chose to leave out any arguments around your
history of contributions as I took your frequency of contributions as a
negative signal on how likely it is I could ping you and get prompt
assistance.

>> Sorry, I misread the description and thought you implied you don't
>> need the light version unless you know why you need it, whereas it is
>> clear now you meant you don't need qemu-traditional unless you know
>> why you need it.
>
> Correct. Feel free to reword that to be clearer.

In this case, I would prefer _you_ make it more clear, but please see
the end of my message.

> Btw, `./nixos/modules/virtualisation/xen-dom0.nix` is fairly crazy too.
> It sets up its own bridges and firewall config, runs its own `dnsmasq`
> and stuff. I kinda planned to fix that later too (currently I just
> comment out that part in my branch and setup everything myself
> per-machine, because who needs options, right?).

Yeah... I know :/ have read it, it is definitely an interesting module.
To be clear, I'm not anti-option. I'm
anti-options-for-the-sake-of-options.

> In short, for me your comment reads like: "more (not useful for me)
> options (I don't use xen, btw) introduce new test cases, meh, don't want
> to run those, let's not".

They are not test cases, they are just packages to compile. This is an
important distinction. Compilation does not guarantee it'll work. Having
more cases may compile but be broken gives a very bad user experience
when a user hits them, and makes it much more complicated to handle when
security patches don't cleanly apply for whatever reason.

> I send the-rays-of-cosmic-love your way for the time you put into
> reviewing this PR, but I disagree.

:) That is okay, we're all pushing for a better NixOS.

> My proposal: read the old expression, read the new expression, ignore
> new configurability, feel the difference (I think the level of space
> required to correctly indent the expression usually correlates with its
> craziness, this patch considerably reduces that metric), then think
> about poor power users that want to run MirageOS on their credit-cards
> with Xen and want all those options, reconsider.

Yeah, I have read the old expression. It is crazy. I do appreciate the
clean-up you have here, it is why I've tried so hard to explain my
position. My goal with this review was:

Don't push back so hard that it is a "no", but do push back hard enough
that it requires real justification on your part.

I think you've made a good case and justification for the work you've
done. In the veign of ffmpeg, have to ask: why offer `-vanilla` and
`-slim` at all? Why not just make our default and only expression the 
`-light` version?

When Xen's XSAs include patches for qemuu and qemut, how should those be
handled? Are those patches fully included in the version of qemu we
already have, and Xen is distributing them only because they have their
own (old?) copy?

Thank you for your thorough reply,
Graham

**Comment by oxij** [comment]
>> The problem with `xen` is that its tarballs are too much
>> "enterprise-level software". Fetching of sources from the net in `make`
>> is nuts, building multiple internal submodules that are lager than `xen`
>> itself is nuts too. Yes, its easier to distribute stuff this way in
>> distribution-agnostic way, but there's no reason to comply with this
>> enterprise-level craziness with nix. Nix can build stuff by itself.
>
> This must already be handled by our current expression, since make can't
> access the network anyway.

Yep, I'm being philosophical there.

>> That's backwards logic: "now we have more test cases, so we have more
>> opportunities to fail them, so lets remove the tests".
>
> No, I disagree. This is adding many more features without test cases.
> Many more ways it could build, but produce invalid results.

I agree that having nixos tests for booting `xen` would be nice, but I'm
not sure if this is possible, I never tried to boot `xen` on top of
`kvm`, and I don't think it will work.

>>> This makes me wonder why someone would want to make these changes in
>>> the first place, and wonder if the additional code complexity is worth
>>> the configurability.
>>
>> Complexity where? All options influence is this:
>
> well, that, but I was also surprised to see the over-all nix expression
> gain 200 lines of code over the old version. Not that this is
> necessarily bad.

Most of that is `.patch` files for things I couldn't fix with sed,
`packages.nix` and doc strings.

>> But, again, if you don't use it yourself then what's the difference? I
>> assume you simply compiled it before. Now you can compile three
>> versions, or, if you're lazy, just continue to compile the vanilla
>> version, I'll take care of the rest. (Actually, the most likely case is
>> that if vanilla compiles, the rest will compile too, since they compile
>> _less_).
>
> I am thrilled to have another Xen maintainer around who I can ping when
> issues come up :)
>
>> Also, if you look at my contribution history to nixpkgs, you'll notice
>> that changes to `xen` were among the first things I ever contributed
>> here af32fd6ce3ce2ae138c2fc2bc5c6067216f8b828
>
> I did notice that, but chose to leave out any arguments around your
> history of contributions as I took your frequency of contributions as a
> negative signal on how likely it is I could ping you and get prompt
> assistance.

Firstly, I had a fair share of rather meh experiences in pushing changes
to nixpkgs when people rejected them for years and then reimplemented
them themselves (with toposorting of `fileSystems` people picked up my
rejected branch #11484, which was nice, but most of the time things
simply get reimplemented, e.g. I'm netbooting NixOS for >3 years now,
and my original rejected proposals got reimplemented in netboot
installer ~30% ATM), most of the time reimplementations are rather 1:1
with my old changes. This is rather demotivating.

Some IMO killer features (like defaults in `nixpkgs.config` for nixos)
were rejected with such fervor that I started my own private fork with
~200 patches ATM which I rebase quite a lot. Since a lot of stuff in
those ~200 patches inter-depends it becomes less and less trivial to
separate pieces that can be merged into vanilla nixpkgs. So most of the
time I do nothing unless stuff gets really annoying to merge (this PR is
an example of that).

Also, most of the time pushing something considerable takes 1x effort in
coding and 5x effort in trying to convince people that they want it.
Which is totally fine, I'm sure I'm like that too from the outside. But
this is exhausting and I'd rather do something else.

Secondly, I hate contexts switches, prefer the "deep work" approach and
so I frequently seclude myself from everything. If you really need my
input, like something is seriously on fire - write me a couple lines
directly, not through github. Direct e-mails have their own filter rules
and I don't ignore those tags for more that a week, usually.

> They are not test cases, they are just packages to compile. This is an

In ideal world it does. 

> important distinction. Compilation does not guarantee it'll work. Having
> more cases may compile but be broken gives a very bad user experience

But there's no way to know if something doesn't work without actually
trying. I doubt more than 10 people use nixpkgs' `xen` ATM. There's no
way we can test everything with such a user base. And without those
options userbase for this expression is gonna be _less_, e.g. I'm not
using vanilla `xen`, so at least one user less.

> when a user hits them, and makes it much more complicated to handle when
> security patches don't cleanly apply for whatever reason.

Yes, but this consideration does not to apply in this case, since
options actually disable internally built stuff.

>> I send the-rays-of-cosmic-love your way for the time you put into
>> reviewing this PR, but I disagree.
>
> :) That is okay, we're all pushing for a better NixOS.

Yay!

>> My proposal: read the old expression, read the new expression, ignore
>> new configurability, feel the difference (I think the level of space
>> required to correctly indent the expression usually correlates with its
>> craziness, this patch considerably reduces that metric), then think
>> about poor power users that want to run MirageOS on their credit-cards
>> with Xen and want all those options, reconsider.
>
> Yeah, I have read the old expression. It is crazy. I do appreciate the
> clean-up you have here, it is why I've tried so hard to explain my
> position. My goal with this review was:
>
> Don't push back so hard that it is a "no", but do push back hard enough
> that it requires real justification on your part.
>
> I think you've made a good case and justification for the work you've
> done. In the veign of ffmpeg, have to ask: why offer `-vanilla` and
> `-slim` at all? Why not just make our default and only expression the 
> `-light` version?

This is the question that bugs me too, yes. As I said before, ideally
`xen` would build only `xen`'s binary and scripts, `qemu` and
`qemu-traditional` would be built with `qemu`'s overriden expression.
But

* the latter is not implemented yet, and I'm not sure `-traditional`
  part is possible,
* and I don't want to break something that seems to work.

(I'm not sure it works because I didn't test it with `xen-dom0.nix`
craziness, nor do I use `xendomains` to run VMs, but I assume it does.)

So, ideally, there would be only `-light` and two on-demand `qemu`s. But
that's not possible right now.

Less ideally, there would be `-light` and `-slim` and single external
`qemu`. That is possible already. But having those two options and
dropping `-vanilla` seems strange, and also not friendly for migrating
users.

Maybe we can switch `xen` to be `xen-light` by default after awhile, but
I'm not comfortable doing it right away. My reasoning: people other than
me will see `-light`, test it in other setups, maybe fix bugs, and then
we could switch.

In short: I'm trying not to break things by doing a minimal change to
the current top-level `xen` derivation.

> When Xen's XSAs include patches for qemuu and qemut, how should those be
> handled? Are those patches fully included in the version of qemu we
> already have, and Xen is distributing them only because they have their
> own (old?) copy?

For now we need to add them into `4.5.nix` as before. But, generally,
yes, `qemuu` is just an old version of top-level `qemu` with security
patches backported.

So, the simplest solution for getting XSAs for top-level `qemu` is to
update its `src`. Alternatively, we can patch it with the same XSA
patches since those XSAs would apply to it too (if it's old enough).

Okay. So what do you want me to do?

**Comment by fpletz** [comment]
First some cc's to people who've updated xen stuff in the past: @ts468 @domenkozar 

Disclaimer: I'm not a xen user, so not going to merge this. Sorry.

> Firstly, I had a fair share of rather meh experiences in pushing changes to nixpkgs when people rejected them for years and then reimplemented them themselves (with toposorting of `fileSystems` people picked up my rejected branch #11484, which was nice, but most of the time things simply get reimplemented, e.g. I'm netbooting NixOS for >3 years now, and my original rejected proposals got reimplemented in netboot
installer ~30% ATM), most of the time reimplementations are rather 1:1 with my old changes. This is rather demotivating.

You are not alone. Even project members can have a hard time getting in useful changes. This is one of the biggest problems we're currently facing and we need a solution soon or we won't be able to scale our development efforts. More people would maintain forks instead of contributing new code. We need a useful and well-defined contribution process but first we need to solve other governance issues in the project,. For instance, we're not able to reach a decision on controversial issues other than "nobody has reverted that commit" or try to invoke our BDFL. This the big think I'd like to tackle in person with project members at the next Nixcon.

Regarding your fork, I can take a look at your work and can the very least merge in stuff that I know about. Do you have it on Github?

> In short: I'm trying not to break things by doing a minimal change to the current top-level `xen` derivation.

Back to the topic of this PR: I think your changes are generally useful and using qemu from nixpkgs is the right choice, if it works. Most xen vulnerabilities seem to be in fact vulnerabilities in qemu, which we would fix anyway.

So from a security standpoint, I would welcome removing the vanilla version or let xen default to xen-light if it doesn't cause any immediate problems for users of the vanilla version (you mentioned some issues but I'm not sure how serious). As 17.03 is already branched off and will be released soon, this could be done for 17.09. Still half a year to go. :smile: 

I agree with @grahamc that we should take care not to needlessly increase the complexity of the expression, but in this case it does seem useful to me.

So :+1: from me to go ahead with this and the other changes you proposed like fixing the xen-dom0 module. Unfortunately, I can't help you much with testing. Thanks for your work!

**Comment by grahamc** [comment]
I'll merge it, though :) thank you for the additional eyes, @fpletz. Thank you again for the contributions, @oxij!

**Comment by grahamc** [comment]
Graham Christensen <notifications@github.com> writes:

> I'll merge it, though :) thank you for the additional eyes, @fpletz. Thank you again for the contributions, @oxij!


I want to give @domenkozar an opportunity to review. I'll merge at
2017-03-08 00:00 CET with or without a review :)

Graham

**Comment by oxij** [comment]
Franz Pletz <notifications@github.com> writes:

> You are not alone. Even project members can have a hard time getting in useful changes. This is one of the biggest problems we're currently facing and we need a solution soon or we won't be able to scale our development efforts. More people would maintain forks instead of contributing new code. We need a useful and well-defined contribution process but first we need to solve other governance issues in the project,. For instance, we're not able to reach a decision on controversial issues other than "nobody has reverted that commit" or try to invoke our BDFL. This the big think I'd like to tackle in person with project members at the next Nixcon.

Good, then let me pour my frustrations here. Feel free to cite,
copy-paste or reuse in any other way anything I wrote here, into other
nixpkgs issues and nix-dev list for that discussion.

IMO one of the worst problems with nixpkgs process is that new hot stuff
gets merged (think systemd, hardening), while reworked for nix tested by
time borrowings from other distros don't.

I know people who now maintain their own forks (we exchange patches
privately from time to time, "nix underground movement", lol) after
hardening got merged in (because of #18995 and/or slowdowns, not
everyone has new shiny Ryzens and Kabby Lakes (or even wants them,
because of Intem ME and AMD PSP)), others switched from NixOS way back
after switch for systemd (and I'm tempted too, every second systemd
update breaks something new. Like recently why why!? did they break
`utmp`? Just why? Now a lot of old useful UNIX stuff doesn't work on
NixOS. Fuck systemd! I'm like super tempted to switch to Gentoo with nix
on top because of that shit, or fork that part of nixpkgs too.).

My fork was started after the original #6968, which did what #12877
discusses the possible implementation of.

The fun fact is: original #6968 can solve #18995 with something like
`nixpkgs.config.stdenv.disableHardining = [ "all" ];` or
`nixpkgs.config.gcc_wrapper.disableHardining = [ "all" ];` in
`configuration.nix`. But `nixpkgs.config` is deprecated and
`packageOverrides` is the way to go, right?

> Regarding your fork, I can take a look at your work and can the very least merge in stuff that I know about. Do you have it on Github?

Nah, it's very personal, because I patch a lot of software to my
preference my branch is a mess.

After #6968 I stopped using overrides. Everything I can do with my
`nixpkgs.config` on steroids I do using it. Everything else I "override"
directly in my nixpkgs fork. This way git produces a merge every time
something potentially breaking the override gets commited to `master`.
Which is way better than debugging `packageOverrides`.

But back to the topic. I'm not sure there's a way to solve controversial
issues like that without actually forking. You can't have
`nixpkgs.config` to be optional, its either ignored and broken/not
useful at all as in `master`, or it informs every package expression in
nixpkgs, as in my fork.

Same with hardening. You either have it on by default, slowdown everyone
and produce #18995, or you don't and you loose some potential benefits
on security breach. Potential because "control flow bending" gets around
hardening, so IMO it's pretty useless, compiling everything with
`-fsanitize=address` would've been more useful, but it's not trivial to
implement and it'll blow the size of binaries, and so that needs a
proper expert discussion (and I don't think I'm enough of an expert for
that, btw).

Or you implement useful `nixpkgs.config` like #6968 did and make
everything configurable. (Note, that `packageOverrides` are not an
option for that, because you don't want to override every single package
by hand.)

But that's a serious policy decision too, because just having a fork
that does that is not useful. You need to actually compile most of the
packages with and without hardening to prevent bitrot and to provide
binary caches.

In my case, I'm just too busy and too lazy to comb my fork and make it
something public can enjoy. But even if I waste some time and do the
combing, what number of people will use it?

I don't think most people have a couple of 8-cores with 32G of RAM
laying around just to compile nixpkgs for themselves. And if you have
less than 12G of RAM then compiling nixpkgs spills into swap (and takes
a week per mass rebuild) or turns into a maintenance job of fixing
compile-time options of every package to reduce RAM requirements. Even
16G is not enough to deterministically compile all of nixpkgs, e.g.
firefox and libreoffice, or libreoffice and Agda start compiling
together and you get OOM.

Also note that our BDFL @edolstra is either on opposite side on every
issue mentioned here, or does not care. Which is kinda depressing.

Also, since GitHub is not a mailing list this issue and discussion will
get forgotten as soon it will get closed. And most people are too lazy
to care for a new issue with "read discussion from #suchandsuch, but
continue that discussion here". (Sometimes I feel like new technology is
conspiring against the older more sane approaches, using network effects
to push better stuff of the market. I can think of a handful of
instances more from the top of my head.)

> Back to the topic of this PR: I think your changes are generally useful and using qemu from nixpkgs is the right choice, if it works. Most xen vulnerabilities seem to be in fact vulnerabilities in qemu, which we would fix anyway.

My thoughts exactly.

> So from a security standpoint, I would welcome removing the vanilla version or let xen default to xen-light if it doesn't cause any immediate problems for users of the vanilla version (you mentioned some issues but I'm not sure how serious). As 17.03 is already branched off and will be released soon, this could be done for 17.09. Still half a year to go. :smile: 

Yep, it has some technical problems, though. Like I have no idea why
`xen-light` needs `seabios` and `ipxe` (which should be external too,
but `xen`'s configure doesn't give such an option ATM).

**Comment by grahamc** [comment]
a few hours late, but thank you for the discussion on this!

**Comment by oxij** [comment]
Thanks, everyone!

**Pull request merged**